{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute equivalence classes based on a gff annotation\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_interval(iv,oivs): # takes current interval and a sorted list of original intervals\n",
    "    res = []\n",
    "    for o in oivs:\n",
    "        if iv[1]>o[0] and o[1]>iv[0]:\n",
    "            res.append(o[2])\n",
    "    return res\n",
    "\n",
    "# TODO: need to use oivs to add information about which transcripts are shared by an equivalence class\n",
    "def process_locus(eps,oivs):\n",
    "    interval = []\n",
    "    res = []\n",
    "    for i,e in enumerate(eps): # iterate over and collect intervals\n",
    "        if len(interval)==2: # full interval found, can be evaluated and reset\n",
    "            segs = evaluate_interval(interval,oivs)\n",
    "            if not len(segs)==0:\n",
    "                res.append(interval+segs)\n",
    "                tmp=interval[1]\n",
    "                if not e[0]==0:\n",
    "                    interval=[tmp,i]\n",
    "                else:\n",
    "                    interval=[tmp]\n",
    "            else:\n",
    "                tmp=interval[1]\n",
    "                if not e[0]==0:\n",
    "                    interval=[tmp,i]\n",
    "                else:\n",
    "                    interval=[tmp]\n",
    "        elif not e[0]==0:\n",
    "            interval.append(i)\n",
    "        else:\n",
    "            continue\n",
    "    if len(interval)==2:\n",
    "        segs = evaluate_interval(interval,oivs)\n",
    "        if not len(segs)==0:\n",
    "            res.append(interval+segs)\n",
    "        \n",
    "    return res\n",
    "\n",
    "def report(segs,outFP,geneID,chrID,strand):\n",
    "    if len(segs)>0:\n",
    "        for s in segs:\n",
    "#             print(s)\n",
    "            outFP.write(\"\\t\".join([chrID,strand,str(s[0]),str(s[1]),str(len(s[2])),geneID,\";\".join(list(s[2]))])+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m--------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-159-d1e288feb358>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     51\u001b[0m                 \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlineCols\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m                 \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlineCols\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m                 \u001b[0mendpoints\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mls\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m                 \u001b[0mendpoints\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mls\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m                 \u001b[0mendpoints\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mls\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "endpoints=[]\n",
    "original_intervals = set() # these contain the transcript boundaries (without separation into exons and introns). \n",
    "# This is necessary in order to preserve splicejunction information whenver possible\n",
    "chrom = \"\"\n",
    "strand = \"+\"\n",
    "ls = 0 # locus start\n",
    "le = 0 # locus end\n",
    "geneID = \"\"\n",
    "ts = 0\n",
    "te = 0\n",
    "transID = \"\"\n",
    "transPar= \"\"\n",
    "\n",
    "inputFP = \"/media/data/genomicData/hg38/annotations/chess2.2.gff\"\n",
    "outFP = \"./chess2.2.ec\"\n",
    "\n",
    "ecFP = open(outFP,\"w+\")\n",
    "\n",
    "with open(inputFP,\"r\") as gff:\n",
    "    for line in gff.readlines():\n",
    "        if(line[0]==\"#\"):\n",
    "            continue\n",
    "        else:\n",
    "            lineCols = line.split(\"\\t\")\n",
    "            if lineCols[2] == \"gene\":\n",
    "                final = process_locus(endpoints,original_intervals) # process the intervals computed for the previous locus\n",
    "                final = [[x[0]+ls,x[1]+ls,tuple(x[2:])] for x in final]\n",
    "                report(final,ecFP,geneID,chrom,strand)\n",
    "                # get some basic information about the locus and initialize things\n",
    "                chrom = lineCols[0]\n",
    "                strand = lineCols[6]\n",
    "                ls = int(lineCols[3])\n",
    "                le = int(lineCols[4])\n",
    "                geneID = lineCols[8].strip().split(\"ID=\")[1].split(\";\")[0]\n",
    "#                 print(geneID)\n",
    "                \n",
    "                endpoints=[[0,set()] for x in range(ls,le+1,1)]\n",
    "                original_intervals=set()\n",
    "                \n",
    "            elif lineCols[2] == \"transcript\":\n",
    "                transID = lineCols[8].strip().split(\"ID=\")[1].split(\";\")[0]\n",
    "                transPar = lineCols[8].strip().split(\"Parent=\")[1].split(\";\")[0]\n",
    "                ts = int(lineCols[3])\n",
    "                te = int(lineCols[4])\n",
    "                \n",
    "                # add to original intervals\n",
    "#                 original_intervals.add((ts-1,te-1,))\n",
    "                \n",
    "            elif lineCols[2] == \"exon\":\n",
    "                parent = lineCols[8].strip().split(\"Parent=\")[1].split(\";\")[0]\n",
    "                start = int(lineCols[3])\n",
    "                end = int(lineCols[4])\n",
    "                endpoints[start-ls][0]+=1\n",
    "                endpoints[start-ls][1].add(parent)\n",
    "                endpoints[end-ls][0]+=1\n",
    "                endpoints[end-ls][1].add(parent)\n",
    "                \n",
    "                # add to original intervals\n",
    "                original_intervals.add((start-ls,end-ls,parent))\n",
    "            else:\n",
    "                continue\n",
    "                \n",
    "    final = process_locus(endpoints,original_intervals)\n",
    "    final = [[x[0]+ls,x[1]+ls,tuple(x[2:])] for x in final]\n",
    "    report(final,outFP,geneID,chrom,strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[10474946, 10474947, ('CHS.360.2', 'CHS.360.1')],\n",
       " [10474947, 10474969, ('CHS.360.2', 'CHS.360.3', 'CHS.360.1')],\n",
       " [10474969, 10474982, ('CHS.360.2', 'CHS.360.3', 'CHS.360.1', 'CHS.360.4')],\n",
       " [10474982,\n",
       "  10475002,\n",
       "  ('CHS.360.2', 'CHS.360.3', 'CHS.360.5', 'CHS.360.1', 'CHS.360.4')],\n",
       " [10475002, 10475080, ('CHS.360.4',)],\n",
       " [10475202, 10475740, ('CHS.360.6',)],\n",
       " [10484499, 10484965, ('CHS.360.7',)],\n",
       " [10484965, 10485115, ('CHS.360.8', 'CHS.360.7')],\n",
       " [10494874, 10495116, ('CHS.360.9',)],\n",
       " [10495274,\n",
       "  10495321,\n",
       "  ('CHS.360.8',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.3',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.5',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.1')],\n",
       " [10521230, 10521231, ('CHS.360.12',)],\n",
       " [10521231, 10521607, ('CHS.360.12', 'CHS.360.13')],\n",
       " [10521607, 10524300, ('CHS.360.13',)],\n",
       " [10524406, 10524483, ('CHS.360.12', 'CHS.360.13')],\n",
       " [10528251, 10528359, ('CHS.360.13',)],\n",
       " [10536005, 10536060, ('CHS.360.14',)],\n",
       " [10536213,\n",
       "  10536297,\n",
       "  ('CHS.360.1',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.3',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.5',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.8')],\n",
       " [10541052, 10549855, ('CHS.360.3',)],\n",
       " [10599238,\n",
       "  10599366,\n",
       "  ('CHS.360.6',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.8',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.5',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.13')],\n",
       " [10605185, 10605603, ('CHS.360.15',)],\n",
       " [10618332,\n",
       "  10618417,\n",
       "  ('CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.5',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.8',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.15',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13')],\n",
       " [10622948, 10623019, ('CHS.360.16',)],\n",
       " [10623019,\n",
       "  10623121,\n",
       "  ('CHS.360.14',\n",
       "   'CHS.360.5',\n",
       "   'CHS.360.8',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.16',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.15')],\n",
       " [10623121, 10623286, ('CHS.360.5',)],\n",
       " [10623895, 10624340, ('CHS.360.17',)],\n",
       " [10624340,\n",
       "  10624410,\n",
       "  ('CHS.360.2',\n",
       "   'CHS.360.8',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.15',\n",
       "   'CHS.360.16',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.17',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12')],\n",
       " [10624410,\n",
       "  10624437,\n",
       "  ('CHS.360.2',\n",
       "   'CHS.360.8',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.18',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.15',\n",
       "   'CHS.360.16',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.17',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12')],\n",
       " [10627272,\n",
       "  10627363,\n",
       "  ('CHS.360.1',\n",
       "   'CHS.360.16',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.17',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.14',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.15',\n",
       "   'CHS.360.18',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.8')],\n",
       " [10627363, 10627439, ('CHS.360.16', 'CHS.360.17')],\n",
       " [10627439, 10627502, ('CHS.360.16',)],\n",
       " [10629531,\n",
       "  10630749,\n",
       "  ('CHS.360.14',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.18',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.15',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.8')],\n",
       " [10630749,\n",
       "  10630757,\n",
       "  ('CHS.360.14',\n",
       "   'CHS.360.4',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.18',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.8')],\n",
       " [10630757,\n",
       "  10630758,\n",
       "  ('CHS.360.14',\n",
       "   'CHS.360.1',\n",
       "   'CHS.360.2',\n",
       "   'CHS.360.13',\n",
       "   'CHS.360.7',\n",
       "   'CHS.360.12',\n",
       "   'CHS.360.18',\n",
       "   'CHS.360.6',\n",
       "   'CHS.360.9',\n",
       "   'CHS.360.8')],\n",
       " [10630758, 10630760, ('CHS.360.14', 'CHS.360.18', 'CHS.360.9')],\n",
       " [10630760, 10630762, ('CHS.360.14', 'CHS.360.9')]]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
